app/controllers/registers_controller.rb

```diff
@@ -57,36 +57,67 @@ def show
+    @entries_with_items = entries.map do |entry_history|
+      current_record = entry_history[:current_entry]
+
+      if entry_history[:previous_entry].nil?
         changed_fields = fields
       else
-        previous_record = records_for_key[previous_record_index]
-        changed_fields = fields.reject { |f| current_record.item.value[f] == previous_record.item.value[f] }
+        previous_record = entry_history[:previous_entry]
+        changed_fields = fields.reject { |f| entry_history[:current_entry].data[f] == previous_record.data[f] }
       end
 
-      { current_record: current_record, previous_record: previous_record, updated_fields: changed_fields, key: entry.key }
+      { current_record: current_record, previous_record: previous_record, updated_fields: changed_fields, key: current_record.key }
     end
 
-    if params[:q].present?
-      filtered = filter(entries_mapped_with_items, params[:q])
-      @entries_with_items = paginate_history(filtered)
-    else
-      @entries_with_items = paginate_history(entries_mapped_with_items)
-    end
+    @entries_with_items = Kaminari::paginate_array(@entries_with_items, total_count: @page_count).page(@current_page).per(100)
   end
 
   private
 
+  def recover_entries_history(register_id, fields, params)
+    literal = params[:q]
+    page = if params[:page].nil? then 1 else params[:page].to_i end
+
+    if literal.present?
+      partial = ''
+      operation_params = []
+      fields.split(',').each { |field| partial += " data->> '#{field}' ilike ? or" }
+      partial = partial[1, partial.length - 3] # Remove beginning space and end 'or'
+
+      operation_params.push(partial)
+      fields.split(',').count.times { operation_params.push("%#{literal}%") }
+
+      query = Entry.where("spina_register_id = '#{register_id}' and entry_type = 'user'").where(operation_params).order(:entry_number).reverse_order.limit(100).offset(100 * (page - 1))
+      count_query = Entry.select(1).where("spina_register_id = '#{register_id}' and entry_type = 'user'").where(operation_params)
+    else
+      query = Entry.where(spina_register_id: register_id, entry_type: 'user').order(:entry_number).reverse_order.limit(100).offset(100 * (page - 1))
+      count_query = Entry.select(1).where(spina_register_id: register_id, entry_type: 'user')
+    end
+
+    previous_entries_numbers = query.reject{ |entry| entry.previous_entry_number.nil? }.map{ |entry| entry.previous_entry_number}
+    previous_entries_query = Entry.where(spina_register_id: register_id, entry_number: previous_entries_numbers)
+
+    result = query.map do |entry|
+      previous_entry = previous_entries_query.select{ |previous_entry| previous_entry.entry_number == entry.previous_entry_number }.first
+      { current_entry: entry, previous_entry: previous_entry }
+    end
+
+    @page_count = count_query.length
+    @current_page = page
+    @total_pages = (@page_count / 100) + (@page_count % 100 == 0 ? 0 : 1)
```

Comment:
think you can use `(@page_count % 100).zero?` here

Reply:
If we're going to change this occurrence then we should change all occurrences (including Javiers).